File: server/reviewservice/src/main/java/com/goodsteams/reviewservice/entity/Review.java
package com.goodsteams.reviewservice.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "review")
public class Review {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "review_id")
    private Long reviewId;

    @Column(name = "book_id", nullable = false)
    private Long bookId;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(name = "rating", nullable = false)
    private Integer rating;

    @Column(name = "body_text", nullable = false, columnDefinition = "TEXT")
    private String bodyText;

    @Column(name = "helpful_count", nullable = false)
    private Integer helpfulCount = 0;

    @Column(name = "not_helpful_count", nullable = false)
    private Integer notHelpfulCount = 0;

    @Column(name = "funny_count", nullable = false)
    private Integer funnyCount = 0;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt = LocalDateTime.now();

    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    @JsonIgnore
    @OneToMany(mappedBy = "review", cascade = CascadeType.ALL, fetch = FetchType.LAZY, orphanRemoval = true)
    private Set<ReviewVote> reviewVotes = new HashSet<>();

    public Review() {}

    public Review(Long bookId, Long userId, int rating, String bodyText) {
        this.bookId = bookId;
        this.userId = userId;
        this.rating = rating;
        this.bodyText = bodyText;
    }

}

File: client/src/types/models/review/Review.ts
export type Review = {
  reviewId: number;
  bookId: number;
  userId: number;
  rating: number;
  bodyText: string;
  helpfulCount: number;
  notHelpfulCount: number;
  funnyCount: number;
  createdAt: string;
  updatedAt: string;
};

export function isReview(obj: unknown): obj is Review {
  const reviewObj = obj as Review;
  return (
    !!obj &&
    typeof reviewObj.reviewId === "number" &&
    typeof reviewObj.bookId === "number" &&
    typeof reviewObj.userId === "number" &&
    typeof reviewObj.rating === "number" &&
    typeof reviewObj.bodyText === "string" &&
    typeof reviewObj.helpfulCount === "number" &&
    typeof reviewObj.notHelpfulCount === "number" &&
    typeof reviewObj.funnyCount === "number" &&
    typeof reviewObj.createdAt === "string" &&
    typeof reviewObj.updatedAt === "string"
  );
}
